### PR TITLE
Fix issues indicated by clippy

### PR DESCRIPTION
**Pull request by neosilky**
This also fixes quite a lot of #139.

Let me know if anything isn't preferred style.

### CODE REVIEW

File: src/datetime.rs [hunk 1]
```diff
@@ -179,11 +179,8 @@ impl FromStr for Datetime {
                               chars.clone().next() == Some('T') {
             chars.next();
             true
-        } else if full_date.is_none() {
-            true
-        } else {
-            false
-        };
+        } else { full_date.is_none() };
```
Comment (alexcrichton) [inline src/datetime.rs]: Can this be spread across multiple lines?

File: src/de.rs [hunk 2]
```diff
@@ -1178,10 +1178,8 @@ impl<'a> Header<'a> {
             }
 
             self.tokens.eat_whitespace()?;
-            if self.require_newline_after_table {
-                if !self.tokens.eat_comment()? {
```
Comment (alexcrichton) [inline src/de.rs]: I prefer this construction, can this remain?

File: src/tokens.rs [hunk 3]
```diff
@@ -66,14 +66,13 @@ impl<'a> Tokenizer<'a> {
...
-        return t
+        t
     }
 
     pub fn next(&mut self) -> Result<Option<Token<'a>>, Error> {
         let token = match self.chars.next() {
             Some((_, '\n')) => Newline,
-            Some((start, ' ')) => self.whitespace_token(start),
-            Some((start, '\t')) => self.whitespace_token(start),
+            Some((start, ' ')) | Some((start, '\t')) => self.whitespace_token(start),
```
Comment (alexcrichton) [inline src/tokens.rs]: I prefer the previous construction, can that remain?

File: src/tokens.rs [hunk 4]
```diff
@@ -101,8 +100,7 @@ impl<'a> Tokenizer<'a> {
     pub fn eat(&mut self, expected: Token<'a>) -> Result<bool, Error> {
         match self.peek()? {
             Some(ref found) if expected == *found => {}
-            Some(_) => return Ok(false),
-            None => return Ok(false),
```
Comment (alexcrichton) [inline src/tokens.rs]: I prefer this construction, can it remain?

File: src/tokens.rs [hunk 5]
```diff
@@ -214,7 +212,7 @@ impl<'a> Tokenizer<'a> {
     }
 
     pub fn current(&mut self) -> usize {
-        self.chars.clone().next().map(|i| i.0).unwrap_or(self.input.len())
+        self.chars.clone().next().map(|i| i.0).unwrap_or_else(|| self.input.len())
```
Comment (alexcrichton) [inline src/tokens.rs]: This change is not necessary

File: src/tokens.rs [hunk 6]
```diff
@@ -229,7 +227,7 @@ impl<'a> Tokenizer<'a> {
     }
 
     fn comment_token(&mut self, start: usize) -> Token<'a> {
-        while let Some((_, ch)) = self.chars.clone().next() {
+        for (_, ch) in self.chars.clone() {
```
Comment (alexcrichton) [inline src/tokens.rs]: This and the one below are semantically different, the current incarnation should remain.

File: tests/valid.rs [hunk 7]
```diff
@@ -47,7 +47,7 @@ fn run(toml: &str, json: &str) {
 
     // Assert toml == json
     let toml_json = to_json(toml.clone());
-    assert!(json == toml_json,
+    assert_eq!(json, toml_json,
```
Comment (alexcrichton) [inline tests/valid.rs]: This isn't necessary

File: src/tokens.rs (hunk 5 above)
Comment (clarcharr) [inline src/tokens.rs]: I filed Manishearth/rust-clippy#1653 for this; there's been a lot of `unwrap_or_else` false positives.

File: src/tokens.rs (hunk 6 above)
Comment (clarcharr) [inline src/tokens.rs]: I filed Manishearth/rust-clippy#1654 for this; this is definitely a bug.

File: tests/valid.rs (hunk 7 above)
Comment (clarcharr) [inline tests/valid.rs]: `assert_eq` actually provides better error messages, unless I'm mistaken?

**Comment by clarcharr** [comment]
There are actually a few false positives here so I've filed clippy bugs for them.

File: tests/valid.rs (hunk 7 above)
Comment (alexcrichton) [inline tests/valid.rs]: Not in this case where the message is specified, no

Comment (clarcharr) [inline tests/valid.rs]: I was under the impression that `assert_eq!` would show the `Debug` for each operand on failure whereas `assert!` would just show the expression `json == toml_json`.

**Comment by neosilky** [comment]
@alexcrichton Sorry about the late reply... Everything should be fixed :+1:

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**